pulp_docker/app/tasks/stages/runner.py

```diff
@@ -0,0 +1,102 @@
+import asyncio
+
+from stages import Stage
+
+
+class ConcurrentRunner(Stage):
+    """
+    Run any Stage concurrently.
+
+    For each item in_q, spin up a new instance of the Stage and run them all concurrently. All
+    output from the concurrent Stage is consolidated into a single out_q.
+    """
+
+    def __init__(self, stage, max_concurrent_content=5):
+        """Initialize the stage."""
+        self.max_concurrent_content = max_concurrent_content
+        self.stage = stage
+
+    @property
+    def saturated(self):
+        """
+        Indicates that max concurrency has been reached.
+        """
+        return len(self.futures) >= self.max_concurrent_content
+
+    async def __call__(self, in_q, out_q):
+        """
+        The coroutine for this stage.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+        Returns:
+            The coroutine for this stage.
+
+        """
+        # TODO add max_concurrent
+        self.futures = set()
+        prev_running = True
+        middle_q = asyncio.Queue()
+        pull_from_queue = asyncio.ensure_future(in_q.get())
+        self.futures.add(pull_from_queue)
+        ready = set()
+        while prev_running or self.futures:
+            done, self.futures = await asyncio.wait(self.futures,
+                                                    return_when=asyncio.FIRST_COMPLETED)
+
+            done = ready.union(done)
+            ready = set()
+            while done:
+                task = done.pop()
+                # if finished task is in_q.get(), start up a new instance of the stage to run
+                # and make a new task for in_q.get()
+                if task is pull_from_queue:
+                    if self.saturated:
+                        ready.add(task)
+                    else:
+                        next_in = task.result()
+                        if next_in is None:
+                            prev_running = False
+                        else:
+                            single_q = SingletonQueue()
+                            await single_q.put(next_in)
+                            self.futures.add(asyncio.ensure_future(self.stage(single_q, middle_q)))
```

Comment:
Given a `Stage` (any object) can be stateful, I don't think it's safe to assume that the same instance can be reused.

Reply:
You are right. I suppose this stage could take uninitialized stages (and the arguments to pass to it, in the form of *args, **kwargs), and create a new instance for each one, however I suspect that might not be memory friendly. The other alternative is to require that stages "wrapped" by this stage are not stateful.